File: implementations/springboot/examples/src/main/java/com/example/orderservice/order/domain/Order.java
package com.example.orderservice.order.domain;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "orders")
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String orderId;

    @Column(nullable = false)
    private String userId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status;

    @ElementCollection
    @CollectionTable(name = "order_items", joinColumns = @JoinColumn(name = "order_id", referencedColumnName = "order_id"))
    private List<OrderItem> items;

    @Column
    private LocalDateTime deletedAt;

    @Transient
    private final List<Object> domainEvents = new ArrayList<>();

    protected Order() {}

    public static Order create(String userId, List<OrderItem> items) {
        if (items == null || items.isEmpty()) {
            throw new IllegalArgumentException("주문 항목은 최소 1개 이상이어야 합니다.");
        }
        Order order = new Order();
        order.orderId = UUID.randomUUID().toString();
        order.userId = userId;
        order.items = new ArrayList<>(items);
        order.status = OrderStatus.PENDING;
        return order;
    }

    public void cancel(String reason) {
        if (this.status == OrderStatus.CANCELLED) throw new IllegalStateException("이미 취소된 주문입니다.");
        if (this.status == OrderStatus.PAID) throw new IllegalStateException("결제 완료된 주문은 취소할 수 없습니다.");
        this.status = OrderStatus.CANCELLED;
        this.domainEvents.add(new OrderCancelledEvent(this.orderId, reason, LocalDateTime.now()));
    }

    public int totalAmount() {
        return items.stream().mapToInt(i -> i.price() * i.quantity()).sum();
    }

    public List<Object> pullDomainEvents() {
        List<Object> events = new ArrayList<>(this.domainEvents);
        this.domainEvents.clear();
        return events;
    }

    public String getOrderId() { return orderId; }
    public String getUserId() { return userId; }
    public OrderStatus getStatus() { return status; }
    public List<OrderItem> getItems() { return List.copyOf(items); }
    public LocalDateTime getDeletedAt() { return deletedAt; }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/orderservice/order/domain/Order.kt
package com.example.orderservice.order.domain

import jakarta.persistence.*
import java.time.LocalDateTime
import java.util.UUID

@Entity
@Table(name = "orders")
class Order protected constructor() {

    @Column(nullable = false, unique = true)
    var orderId: String = ""
        private set

    @Column(nullable = false)
    var userId: String = ""
        private set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: OrderStatus = OrderStatus.PENDING
        private set

    @ElementCollection
    @CollectionTable(
        name = "order_items",
        joinColumns = [JoinColumn(name = "order_id", referencedColumnName = "order_id")]
    )
    var items: MutableList<OrderItem> = mutableListOf()
        private set

    @Column
    var deletedAt: LocalDateTime? = null
        private set

    @Transient
    private val domainEvents: MutableList<Any> = mutableListOf()

    companion object {
        fun create(userId: String, items: List<OrderItem>): Order {
            require(items.isNotEmpty()) { "주문 항목은 최소 1개 이상이어야 합니다." }
            return Order().apply {
                this.orderId = UUID.randomUUID().toString()
                this.userId = userId
                this.items = items.toMutableList()
                this.status = OrderStatus.PENDING
            }
        }
    }

    fun cancel(reason: String) {
        if (status == OrderStatus.CANCELLED) throw OrderAlreadyCancelledException()
        if (status == OrderStatus.PAID) throw OrderPaidNotCancellableException()
        status = OrderStatus.CANCELLED
        domainEvents += OrderCancelledEvent(orderId, reason, LocalDateTime.now())
    }

    fun totalAmount(): Int = items.sumOf { it.price * it.quantity }

    fun pullDomainEvents(): List<Any> = domainEvents.toList().also { domainEvents.clear() }
}
